src/core/internal/arrayop.d

```diff
@@ -0,0 +1,453 @@
+module core.internal.arrayop;
+import core.internal.traits : Filter, Unqual;
+
+version (X86)
+{
+    version (DigitalMars)
+        version = NoSIMD; // DMD-x86 does not support SIMD
+    else version = X86_OR_X64;
+}
+else version (X86_64)
+    version = X86_OR_X64;
+else
+    static assert(0, "unimplemented");
+
+version (GNU) version = GNU_OR_LDC;
+version (LDC) version = GNU_OR_LDC;
+
+T[] arrayOp(T : T[], Args...)(T[] res, Filter!(isType, Args) args) @trusted @nogc pure nothrow
+{
+    enum check = opsSupported!(true, T, Filter!(not!isType, Args)); // must support all scalar ops
+
+    size_t pos;
+    static if (vectorizeable!(T[], Args))
+    {
+        alias vec = .vec!T;
+        alias load = .load!(T, vec.length);
+        alias store = .store!(T, vec.length);
+
+        // Given that there are at most as many scalars broadcast as there are
+        // operations in any `ary[] = ary[] op const op const`, it should always be
+        // worthwhile to choose vector operations.
+        if (res.length >= vec.length)
+        {
+            mixin(initScalarVecs!Args);
+
+            auto n = res.length / vec.length;
+            do
+            {
+                mixin(vectorExp!Args ~ ";");
+                pos += vec.length;
+            }
+            while (--n);
+        }
+    }
+    for (; pos < res.length; ++pos)
+        mixin(scalarExp!Args ~ ";");
```

Comment:
Two really important pieces of information that could be most beneficial are:
1. Was this instantiated in only by CTFE? We never need these.
2. Was this instantiated inside a function or module/class scope?  With instantiations from a function it should be fine to discard inlined and unreferenced functions in this current compilation.